pex/pex_bootstrapper.py

```diff
@@ -104,36 +115,51 @@ def maybe_reexec_pex(compatibility_constraints):
   :param compatibility_constraints: list of requirements-style strings that constrain the
   Python interpreter to re-exec this pex with.
   """
-  if os.environ.pop('SHOULD_EXIT_BOOTSTRAP_REEXEC', None):
+
+  current_interpreter = PythonInterpreter.get()
+
+  current_interpreter_blessed_env_var = '_PEX_SHOULD_EXIT_BOOTSTRAP_REEXEC'
+  if os.environ.pop(current_interpreter_blessed_env_var, None):
     # We've already been here and selected an interpreter. Continue to execution.
     return
 
-  target = None
-  with TRACER.timed('Selecting runtime interpreter based on pexrc', V=3):
+  compatibility_constraints = pex_info.interpreter_constraints
+  with TRACER.timed('Selecting runtime interpreter', V=3):
     if ENV.PEX_PYTHON and not ENV.PEX_PYTHON_PATH:
       # preserve PEX_PYTHON re-exec for backwards compatibility
       # TODO: Kill this off completely in favor of PEX_PYTHON_PATH
       # https://github.com/pantsbuild/pex/issues/431
+      TRACER.log('Using PEX_PYTHON={} constrained by {}'
+                 .format(ENV.PEX_PYTHON, compatibility_constraints), V=3)
       target = _select_pex_python_interpreter(ENV.PEX_PYTHON,
                                               compatibility_constraints=compatibility_constraints)
-    elif ENV.PEX_PYTHON_PATH:
-      target = _select_interpreter(pex_python_path=ENV.PEX_PYTHON_PATH,
-                                   compatibility_constraints=compatibility_constraints)
-
-    elif compatibility_constraints:
-      # Apply constraints to target using regular PATH
-      target = _select_interpreter(compatibility_constraints=compatibility_constraints)
-
-  if target and os.path.realpath(target) != os.path.realpath(sys.executable):
-    cmdline = [target] + sys.argv
-    TRACER.log('Re-executing: cmdline="%s", sys.executable="%s", PEX_PYTHON="%s", '
-               'PEX_PYTHON_PATH="%s", COMPATIBILITY_CONSTRAINTS="%s"'
-               % (cmdline, sys.executable, ENV.PEX_PYTHON, ENV.PEX_PYTHON_PATH,
-                  compatibility_constraints))
-    ENV.delete('PEX_PYTHON')
-    ENV.delete('PEX_PYTHON_PATH')
-    os.environ['SHOULD_EXIT_BOOTSTRAP_REEXEC'] = '1'
-    os.execve(target, cmdline, ENV.copy())
+    elif ENV.PEX_PYTHON_PATH or compatibility_constraints:
+      TRACER.log('Using PEX_PYTHON_PATH={} constrained by {}'
+                 .format(ENV.PEX_PYTHON, compatibility_constraints), V=3)
```

Comment:
Should be `ENV.PEX_PYTHON_PATH`.

Reply:
Also this logging statement could be confusing if `PEX_PYTHON_PATH` is `None`. The message could be based against a check:

```python
if ENV.PEX_PYTHON_PATH is not None:
   log('Using PEX_PYTHON_PATH={}, constrained by {}')
else:
  log('Using your $PATH, constrained by {}')
```

Reply:
Fixed.